packages/@aws-cdk/aws-synthetics/lib/code.ts

```diff
@@ -0,0 +1,177 @@
+   * assume it's initialized. You may just use it as a construct scope.
+   */
+  public abstract bind(scope: Construct, handler: string): CodeConfig;
+}
+
+/**
+ * Configuration of the code class
+ */
+export interface CodeConfig {
+  /**
+   * The location of the code in S3 (mutually exclusive with `inlineCode`).
+   *
+   * @default - mutually exclusive with `inlineCode`
+   */
+  readonly s3Location?: s3.Location;
+
+  /**
+   * Inline code (mutually exclusive with `s3Location`).
+   *
+   * @default - mutually exclusive with `s3Location`
+   */
+  readonly inlineCode?: string;
+}
+
+/**
+ * Canary code from an Asset
+ */
+export class AssetCode extends Code {
+  private asset?: s3_assets.Asset;
+
+  /**
+   * @param assetPath The path to the asset file or directory.
+   */
+  constructor(private assetPath: string, private options?: s3_assets.AssetOptions){
+    super();
+  }
+
+  public bind(scope: Construct, handler: string): CodeConfig {
+    this.validateCanaryAsset(handler);
+
+    // If the same AssetCode is used multiple times, retain only the first instantiation.
+    if (!this.asset){
+      this.asset = new s3_assets.Asset(scope, 'Code', {
+        path: this.assetPath,
+        ...this.options,
+      });
+    }
+
+    return {
+      s3Location: {
+        bucketName: this.asset.s3BucketName,
+        objectKey: this.asset.s3ObjectKey,
+      },
+    };
+  }
+
+  /**
+   * Validates requirements specified by the canary resource. For example, the canary code with handler `index.handler`
+   * must be found in the file structure `nodejs/node_modules/index.js`.
+   *
+   * Requires path to be either zip file or directory.
+   * Requires asset directory to have the structure 'nodejs/node_modules'.
+   * Requires canary file to be directly inside node_modules folder.
+   * Requires canary file name matches the handler name.
+   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html
+   *
+   * @param handler the canary handler
+   */
+  private validateCanaryAsset(handler: string){
+    if(path.extname(this.assetPath) !== '.zip'){
+      if(path.extname(this.assetPath) !== ''){
+        throw new Error(`Asset must be a .zip file or a directory (${this.assetPath})`);
+      }
+      const filename = handler.substring(0,handler.indexOf('.')) + '.js';
+      if(!fs.existsSync(path.join(this.assetPath,'nodejs', 'node_modules', filename))){
+        throw new Error(`The canary resource requires that the directory contains nodejs${path.sep}node_modules${path.sep}<filename> and that <filename> matches the handler (${filename})`);
+      }
+    }
+  }
+}
+
+/**
+ * Canary code from an inline string (limited to 4KiB).
+ */
+export class InlineCode extends Code {
+  constructor(private code: string) {
+    super();
+
+    if (code.length === 0) {
+      throw new Error('Canary inline code cannot be empty');
+    }
+
+    if (code.length > 460800) {
```

Comment:
This is supposed to be 460KiB (and the 4KiB message was wrongly copied from Lambda which is 100% my bad).

Side note: the [CFN docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-synthetics-canary-code.html#aws-properties-synthetics-canary-code-properties) say up to 5 MB for an inline script, which I've been told is ludicrous. 

The 460800 bytes restriction is the error I get from CFN when testing this out. Will this do what I want? I'm not sure what the `_` does.

```ts
if (code.length > 460_800) {
```

Reply:
So I've dropped this requirement. It seems like the 460800 byte restriction is the typical CFN restriction for stacks rather than a canary restriction. As for the canary restriction, its "5 MB" according to the docs, but that doesn't come into play due to the CFN restriction.